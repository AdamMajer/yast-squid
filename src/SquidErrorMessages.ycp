/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/SquidErrorMessages.ycp
 * Package:	Configuration of squid
 * Summary: Handle list of paths to direcotories with error messages
 *          in different languages.
 * Authors:	Daniel Fiser <dfiser@suse.cz>
 *
 * $Id$
 *
 */

{

module "SquidErrorMessages";
textdomain "squid";

/**
 * format:
 *      $[ "language" : "path_to_directory",
 *         ....
 *       ]
 */
map<string, string> err = $[];

string err_msg_dir = "/usr/share/squid/errors";


void read()
{
    //if err uninitialized
    if (size(err) == 0){
        map<string, any> bash_output = (map<string, any>)SCR::Execute(.target.bash_output, "cd " + err_msg_dir + "; ls -d */");
        err = $[];
        foreach (string dir, splitstring(bash_output["stdout"]:"","\n"),
            {
                dir = deletechars(dir, "/");
                dir = mergestring(splitstring(dir, "_"), " ");
                if (size(dir)>0)
                    err[dir] = err_msg_dir + "/" + dir;
            });

        y2debug("SquidErrorMessages::read() - err: %1", err);
    }
}

/**
 * Returns list of all available languages
 */
global list<string> GetLanguages()
{
    read();

    list<string> ret = [];
    foreach (string key, any value, err,
        {
            ret = add(ret, key);
        });
    return ret;
}

/**
 * Returns list of all available languages in form of items of ComboBox.
 */
global list<term> GetLanguagesToComboBox()
{
    read();

    list<term> ret = [];
    foreach (string language, GetLanguages(),
        {
            ret = add(ret, `item(`id(language), language));
        });
    return ret;
}


/**
 * Returns path to directory containing error messages in given language.
 */
global string GetPath(string language)
{
    return err[language]:"";
}

/**
 * Inverse function to GetPath.
 * Returns languge which has path pth to directory containing error messages.
 */
global string GetLanguageFromPath(string pth)
{
    string ret = nil;
    foreach (string key, string value, err,
        {
            if (value == pth){
                ret = key;
                break;
            }
        });

    return ret;
}

}
/* vim: set sw=4 ts=4 et ft=ycp : */
