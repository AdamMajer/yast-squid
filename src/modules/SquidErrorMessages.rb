# encoding: utf-8

# ------------------------------------------------------------------------------
# Copyright (c) 2006 Novell, Inc. All Rights Reserved.
#
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of version 2 of the GNU General Public License as published by the
# Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail, you may find
# current contact information at www.novell.com.
# ------------------------------------------------------------------------------

# File:	modules/SquidErrorMessages.ycp
# Package:	Configuration of squid
# Summary: Handle list of paths to direcotories with error messages
#          in different languages.
# Authors:	Daniel Fiser <dfiser@suse.cz>
#
# $Id$
#
require "yast"

module Yast
  class SquidErrorMessagesClass < Module
    def main
      textdomain "squid"

      Yast.import "FileUtils"


      # format:
      #      $[ "language" : "path_to_directory",
      #         ....
      #       ]
      @err = {}

      # Directory where is located directories with messages in various languages.
      @err_msg_dir = "/usr/share/squid/errors"

      @trans_map = {
        # language name - combo box entry
        "Armenian"            => _("Armenian"),
        # language name - combo box entry
        "Catalan"             => _("Catalan"),
        # language name - combo box entry
        "Dutch"               => _("Dutch"),
        # language name - combo box entry
        "Finnish"             => _("Finnish"),
        # language name - combo box entry
        "Greek"               => _("Greek"),
        # language name - combo box entry
        "Italian"             => _("Italian"),
        # language name - combo box entry
        "Lithuanian"          => _(
          "Lithuanian"
        ),
        # language name - combo box entry
        "Romanian"            => _("Romanian"),
        # language name - combo box entry
        "Serbian"             => _("Serbian"),
        # language name - combo box entry
        "Spanish"             => _("Spanish"),
        # language name - combo box entry
        "Turkish"             => _("Turkish"),
        # language name - combo box entry; don't translate the encoding suffix
        "Ukrainian-1251"      => _(
          "Ukrainian-1251"
        ),
        # language name - combo box entry
        "Azerbaijani"         => _(
          "Azerbaijani"
        ),
        # language name - combo box entry
        "Czech"               => _("Czech"),
        # language name - combo box entry
        "English"             => _("English"),
        # language name - combo box entry
        "French"              => _("French"),
        # language name - combo box entry
        "Hebrew"              => _("Hebrew"),
        # language name - combo box entry
        "Japanese"            => _("Japanese"),
        # language name - combo box entry
        "Polish"              => _("Polish"),
        # language name - combo box entry; don't translate the encoding suffix
        "Russian-koi8-r"      => _(
          "Russian-koi8-r"
        ),
        # language name - combo box entry
        "Simplify Chinese"    => _(
          "Simplified Chinese"
        ),
        # language name - combo box entry
        "Swedish"             => _("Swedish"),
        # language name - combo box entry; don't translate the encoding suffix
        "Ukrainian-koi8-u"    => _(
          "Ukrainian-koi8-u"
        ),
        # language name - combo box entry
        "Bulgarian"           => _(
          "Bulgarian"
        ),
        # language name - combo box entry
        "Danish"              => _("Danish"),
        # language name - combo box entry
        "Estonian"            => _("Estonian"),
        # language name - combo box entry
        "German"              => _("German"),
        # language name - combo box entry
        "Hungarian"           => _(
          "Hungarian"
        ),
        # language name - combo box entry
        "Korean"              => _("Korean"),
        # language name - combo box entry
        "Portuguese"          => _(
          "Portuguese"
        ),
        # language name - combo box entry; don't translate the encoding suffix
        "Russian-1251"        => _(
          "Russian-1251"
        ),
        # language name - combo box entry
        "Slovak"              => _("Slovak"),
        # language name - combo box entry
        "Traditional Chinese" => _(
          "Traditional Chinese"
        ),
        # language name - combo box entry; don't translate the encoding suffix
        "Ukrainian-utf8"      => _(
          "Ukrainian-utf8"
        )
      }
    end

    def read
      #if err uninitialized else do nothing
      if Builtins.size(@err) == 0
        @err = {}
        dir = ""
        Builtins.foreach(
          Convert.convert(
            SCR.Read(path(".target.dir"), @err_msg_dir),
            :from => "any",
            :to   => "list <string>"
          )
        ) do |value|
          if FileUtils.IsDirectory(Ops.add(Ops.add(@err_msg_dir, "/"), value))
            dir = Builtins.mergestring(Builtins.splitstring(value, "_"), " ")
            if Ops.greater_than(Builtins.size(dir), 0)
              Ops.set(@err, dir, Ops.add(Ops.add(@err_msg_dir, "/"), value))
            end
          end
        end

        Builtins.y2debug("SquidErrorMessages::read() - err: %1", @err)
      end

      nil
    end

    # Returns list of all available languages
    def GetLanguages
      read

      ret = []
      Builtins.foreach(@err) { |key, value| ret = Builtins.add(ret, key) }
      deep_copy(ret)
    end

    # Returns list of all available languages in form of items of ComboBox.
    def GetLanguagesToComboBox
      read

      ret = []
      Builtins.foreach(GetLanguages()) do |language|
        ret = Builtins.add(
          ret,
          Item(Id(language), Ops.get(@trans_map, language, language))
        )
      end
      deep_copy(ret)
    end


    # Returns path to directory containing error messages in given language.
    def GetPath(language)
      read

      Ops.get(@err, language, "")
    end

    # Inverse function to GetPath.
    # Returns languge which has path pth to directory containing error messages.
    def GetLanguageFromPath(pth)
      read

      ret = nil
      Builtins.foreach(@err) do |key, value|
        if value == pth
          ret = key
          raise Break
        end
      end

      ret
    end

    publish :variable => :err, :type => "map <string, string>", :private => true
    publish :variable => :err_msg_dir, :type => "string", :private => true
    publish :function => :read, :type => "void ()", :private => true
    publish :function => :GetLanguages, :type => "list <string> ()"
    publish :variable => :trans_map, :type => "map <string, string>", :private => true
    publish :function => :GetLanguagesToComboBox, :type => "list <term> ()"
    publish :function => :GetPath, :type => "string (string)"
    publish :function => :GetLanguageFromPath, :type => "string (string)"
  end

  SquidErrorMessages = SquidErrorMessagesClass.new
  SquidErrorMessages.main
end
