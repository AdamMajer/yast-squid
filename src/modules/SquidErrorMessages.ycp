/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/SquidErrorMessages.ycp
 * Package:	Configuration of squid
 * Summary: Handle list of paths to direcotories with error messages
 *          in different languages.
 * Authors:	Daniel Fiser <dfiser@suse.cz>
 *
 * $Id$
 *
 */

{

module "SquidErrorMessages";
textdomain "squid";

import "FileUtils";


/**
 * format:
 *      $[ "language" : "path_to_directory",
 *         ....
 *       ]
 */
map<string, string> err = $[];

/**
 * Directory where is located directories with messages in various languages.
 */
string err_msg_dir = "/usr/share/squid/errors";


void read()
{
    //if err uninitialized else do nothing
    if (size(err) == 0){
        err = $[];
        string dir = "";
        foreach (string value, (list<string>)SCR::Read(.target.dir, err_msg_dir),
            {
                if (FileUtils::IsDirectory(err_msg_dir + "/" + value)){
                    dir = mergestring(splitstring(value, "_"), " ");
                    if (size(dir) > 0){
                        err[dir] = err_msg_dir + "/" + value;
                    }
                }
            });

        y2debug("SquidErrorMessages::read() - err: %1", err);
    }
}

/**
 * Returns list of all available languages
 */
global list<string> GetLanguages()
{
    read();

    list<string> ret = [];
    foreach (string key, any value, err,
        {
            ret = add(ret, key);
        });
    return ret;
}

map<string,string> trans_map = $[
    // language name - combo box entry
    "Armenian" : _("Armenian"),
    // language name - combo box entry
    "Catalan" : _("Catalan"),
    // language name - combo box entry
    "Dutch" : _("Dutch"),
    // language name - combo box entry
    "Finnish" : _("Finnish"),
    // language name - combo box entry
    "Greek" : _("Greek"),
    // language name - combo box entry
    "Italian" : _("Italian"),
    // language name - combo box entry
    "Lithuanian" : _("Lithuanian"),
    // language name - combo box entry
    "Romanian" : _("Romanian"),
    // language name - combo box entry
    "Serbian" : _("Serbian"),
    // language name - combo box entry
    "Spanish" : _("Spanish"),
    // language name - combo box entry
    "Turkish" : _("Turkish"),
    // language name - combo box entry; don't translate the encoding suffix
    "Ukrainian-1251" : _("Ukrainian-1251"),
    // language name - combo box entry
    "Azerbaijani" : _("Azerbaijani"),
    // language name - combo box entry
    "Czech" : _("Czech"),
    // language name - combo box entry
    "English" : _("English"),
    // language name - combo box entry
    "French" : _("French"),
    // language name - combo box entry
    "Hebrew" : _("Hebrew"),
    // language name - combo box entry
    "Japanese" : _("Japanese"),
    // language name - combo box entry
    "Polish" : _("Polish"),
    // language name - combo box entry; don't translate the encoding suffix
    "Russian-koi8-r" : _("Russian-koi8-r"),
    // language name - combo box entry
    "Simplify Chinese" : _("Simplified Chinese"),
    // language name - combo box entry
    "Swedish" : _("Swedish"),
    // language name - combo box entry; don't translate the encoding suffix
    "Ukrainian-koi8-u" : _("Ukrainian-koi8-u"),
    // language name - combo box entry
    "Bulgarian" : _("Bulgarian"),
    // language name - combo box entry
    "Danish" : _("Danish"),
    // language name - combo box entry
    "Estonian" : _("Estonian"),
    // language name - combo box entry
    "German" : _("German"),
    // language name - combo box entry
    "Hungarian" : _("Hungarian"),
    // language name - combo box entry
    "Korean" : _("Korean"),
    // language name - combo box entry
    "Portuguese" : _("Portuguese"),
    // language name - combo box entry; don't translate the encoding suffix
    "Russian-1251" : _("Russian-1251"),
    // language name - combo box entry
    "Slovak" : _("Slovak"),
    // language name - combo box entry
    "Traditional Chinese" : _("Traditional Chinese"),
    // language name - combo box entry; don't translate the encoding suffix
    "Ukrainian-utf8" : _("Ukrainian-utf8"),
];

/**
 * Returns list of all available languages in form of items of ComboBox.
 */
global list<term> GetLanguagesToComboBox()
{
    read();

    list<term> ret = [];
    foreach (string language, GetLanguages(),
        {
            ret = add(ret, `item(`id(language), trans_map[language]:language));
        });
    return ret;
}


/**
 * Returns path to directory containing error messages in given language.
 */
global string GetPath(string language)
{
    read();

    return err[language]:"";
}

/**
 * Inverse function to GetPath.
 * Returns languge which has path pth to directory containing error messages.
 */
global string GetLanguageFromPath(string pth)
{
    read();

    string ret = nil;
    foreach (string key, string value, err,
        {
            if (value == pth){
                ret = key;
                break;
            }
        });

    return ret;
}

}
/* vim: set sw=4 ts=4 et ft=ycp : */
