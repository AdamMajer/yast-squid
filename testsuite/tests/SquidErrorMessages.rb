# encoding: utf-8

module Yast
  class SquidErrorMessagesClient < Client
    def main
      # testedfiles: SquidErrorMessages.ycp

      @READ = [
        {
          "target" => {
            "dir" => [
              "Russian-1251",
              "a_file",
              "English",
              "b_file",
              "Simplify_Chinese",
              "c_file"
            ]
          }
        },
        { "target" => { "stat" => { "isdir" => true } } },
        { "target" => { "stat" => { "isdir" => false } } },
        { "target" => { "stat" => { "isdir" => true } } },
        { "target" => { "stat" => { "isdir" => false } } },
        { "target" => { "stat" => { "isdir" => true } } },
        { "target" => { "stat" => { "isdir" => false } } }
      ]

      @WRITE = {}
      @EXECUTE = {}

      Yast.include self, "testsuite.rb"

      Yast.import "SquidErrorMessages"
      Yast.import "FileUtils"


      DUMP("GetLanguages()")
      TEST(lambda { SquidErrorMessages.GetLanguages }, [@READ, @WRITE, @EXECUTE], nil)

      DUMP("")
      DUMP("GetLanguagesToComboBox()")
      TEST(lambda { SquidErrorMessages.GetLanguagesToComboBox }, [
        @READ,
        @WRITE,
        @EXECUTE
      ], nil)

      DUMP("")
      DUMP("GetPath(\"English\")")
      TEST(lambda { SquidErrorMessages.GetPath("English") }, [
        @READ,
        @WRITE,
        @EXECUTE
      ], nil)
      DUMP("GetPath(\"Simplify Chinese\")")
      TEST(lambda { SquidErrorMessages.GetPath("Simplify Chinese") }, [
        @READ,
        @WRITE,
        @EXECUTE
      ], nil)

      DUMP("")
      DUMP(
        Ops.add(
          Ops.add(
            Ops.add("GetLanguageFromPath(", SquidErrorMessages.err_msg_dir),
            "/English"
          ),
          ")"
        )
      )
      TEST(lambda do
        SquidErrorMessages.GetLanguageFromPath(
          Ops.add(SquidErrorMessages.err_msg_dir, "/English")
        )
      end, [
        @READ,
        @WRITE,
        @EXECUTE
      ], nil)
      DUMP(
        Ops.add(
          Ops.add(
            Ops.add("GetLanguageFromPath(", SquidErrorMessages.err_msg_dir),
            "/Simplify_Chinese"
          ),
          ")"
        )
      )
      TEST(lambda do
        SquidErrorMessages.GetLanguageFromPath(
          Ops.add(SquidErrorMessages.err_msg_dir, "/Simplify_Chinese")
        )
      end, [
        @READ,
        @WRITE,
        @EXECUTE
      ], nil)

      nil
    end
  end
end

Yast::SquidErrorMessagesClient.new.main
