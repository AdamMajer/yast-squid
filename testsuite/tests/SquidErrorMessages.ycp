{

// testedfiles: SquidErrorMessages.ycp

any READ = [
    $[ "target" : $[
        "dir" : [ "Russian-1251",
                  "a_file",
                  "English",
                  "b_file",
                  "Simplify_Chinese",
                  "c_file"]
    ]],

    $[ "target" : $[ "stat" : $[ "isdir" : true]]],
    $[ "target" : $[ "stat" : $[ "isdir" : false]]],
    $[ "target" : $[ "stat" : $[ "isdir" : true]]],
    $[ "target" : $[ "stat" : $[ "isdir" : false]]],
    $[ "target" : $[ "stat" : $[ "isdir" : true]]],
    $[ "target" : $[ "stat" : $[ "isdir" : false]]],
];

map WRITE = $[];
map EXECUTE = $[];

include "testsuite.ycp";

import "SquidErrorMessages";
import "FileUtils";


DUMP("GetLanguages()");
TEST(``(SquidErrorMessages::GetLanguages()),[READ,WRITE,EXECUTE], nil);

DUMP("");
DUMP("GetLanguagesToComboBox()");
TEST(``(SquidErrorMessages::GetLanguagesToComboBox()),[READ,WRITE,EXECUTE], nil);

DUMP("");
DUMP("GetPath(\"English\")");
TEST(``(SquidErrorMessages::GetPath("English")),[READ,WRITE,EXECUTE], nil);
DUMP("GetPath(\"Simplify Chinese\")");
TEST(``(SquidErrorMessages::GetPath("Simplify Chinese")),[READ,WRITE,EXECUTE], nil);

DUMP("");
DUMP("GetLanguageFromPath(" + SquidErrorMessages::err_msg_dir + "/English" + ")");
TEST(``(SquidErrorMessages::GetLanguageFromPath(SquidErrorMessages::err_msg_dir + "/English")),[READ,WRITE,EXECUTE], nil);
DUMP("GetLanguageFromPath(" + SquidErrorMessages::err_msg_dir + "/Simplify_Chinese" + ")");
TEST(``(SquidErrorMessages::GetLanguageFromPath(SquidErrorMessages::err_msg_dir + "/Simplify_Chinese")),[READ,WRITE,EXECUTE], nil);

}
/* vim: set sw=4 ts=4 et ft=ycp : */
