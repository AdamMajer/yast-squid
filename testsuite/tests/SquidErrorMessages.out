Dump	GetLanguages()
Read	.target.dir "/usr/share/squid/errors" ["en", "a_file", "ru", "b_file", "uk", "c_file"]
Return	["en", "ru", "uk"]
Dump	
Dump	GetLanguagesToComboBox()
Return	[`item (`id ("en"), "English"), `item (`id ("ru"), "Russian"), `item (`id ("uk"), "Ukrainian")]
Dump	
Dump	GetPath("en")
Return	/usr/share/squid/errors/en
Dump	GetPath("uk")
Return	/usr/share/squid/errors/uk
Dump	
Dump	GetLanguageFromPath(/usr/share/squid/errors/en)
Return	en
Dump	GetLanguageFromPath(/usr/share/squid/errors/uk)
Return	ru
